### DISCUSSION TITLE
Long targets pipeline

### DISCUSSION THREAD

**Discussion by tolliam** [ropensci/targets#850] [Help]
Has anyone ran in to difficultly of an R {targets} pipeline script getting too long and cumbersome?
The solutions I've come up with are:
•	Getting rid of old targets (not so good for reproducibility)
•	Starting again with new project/pipeline
•	Putting archive {targets} in a child script that gets sourced from main _targets.R script (I've tried this and it seems to work ok)
•	Writing wrapper functions for any new models that hardcode within the function any parameter settings (to prevent script getting ever longer)
Any thoughts?

**Comment by wlandau** [comment]
You can also modularize the target script. For example, if you have:

```r
# _targets.R
library(targets)
source("R/functions.R")
list(
  tar_target(data1, ...),
  tar_target(data2, ...),
  tar_target(data3, ...),
  tar_target(model1, ...),
  tar_target(model2, ...),
  tar_target(model3, ...)
)
```

You can condense it down into 

```r
# _targets.R
library(targets)
source("R/functions.R")
source("R/targets_data.R")
source("R/targets_model.R")
list( # a shorter list
  targets_data,
  targets_model
)
```

where

```r
# R/targets_data.R
targets_data <- list(
  tar_target(data1, ...),
  tar_target(data2, ...),
  tar_target(data3, ...)
)
```

and 

```r
# R/targets_model.R
targets_model <- list(
  tar_target(model1, ...),
  tar_target(model2, ...),
  tar_target(model3, ...)
)
```

**Reply by wlandau** [reply]
Functions also shorten the target list by shortening each target definition, but it sounds like you are already doing that.

**Reply by tolliam** [reply]
Thanks @wlandau, that is similar to the sort of workflow I was just thinking of earlier but better. Rather than archive some old targets in to a child script, take this opportunity to put all the targets in to child scripts/lists relevant to the different parts of the modelling. Many thanks!

**Comment by benzipperer** [comment]
Making the target script more modular is very helpful for me, but I'm having trouble understanding why the objects created in the separate files remain outdated?  Here is a simple example:

```r
lines <- c(
  "targets_models <- list(",
  "tar_target(model1, lm(mpg ~ wt, data = mtcars))",
  ")"
)

dir.create("R")
writeLines(lines, "R/targets_models.R") # In tar_dir(), not the user's file space.

targets::tar_script({
  source("R/targets_models.R")
  list(
    targets_models
  )
}, ask = FALSE)
```

After running the pipeline, why does targets_models remain outdated? 

```r
targets::tar_outdated(targets_only = F)
#> [1] "targets_models" "model1"

targets::tar_make()
#> • start target model1
#> • built target model1 [0.002 seconds]
#> • end pipeline [0.04 seconds]

targets::tar_outdated(targets_only = F)
#> [1] "targets_models"
```

**Reply by wlandau** [reply]
You do not need to worry about `targets_models`. That one is a target specification object whose hash changes unpredictably from R session to R session. You use it to *specify* the targets you are going to run, but it is not a data object for any of the return values of any of the targets. If it shows up as outdated, you can ignore it.

**Reply by wlandau** [reply]
The critical things to watch for are actually target names like `model1`.
